Fix for FAB-17149,FAB-17151,FAB-17152

SamYuan1990 opened the pull request:
Signed-off-by: SamYuan1990 yy19902439@126.com
Type of change
Bug fix

Description
Fix for FAB-17151, FAB-17152, FAB-17149
1) Rename test folder name in new format.
2) Updated clean up function cleaned temp dir created in test cases.
3) Updated test case to invoke cleanup function with defer to remove temp files.

Additional details
To verify locally, cd core/ledger/kvledger and go test -v ./... then ls test-env

Related issues
https://jira.hyperledger.org/browse/FAB-17149
https://jira.hyperledger.org/browse/FAB-17151
https://jira.hyperledger.org/browse/FAB-17152

SamYuan1990 committed:
Fix for FAB-17151, FAB-17152, FAB-17149

Signed-off-by: SamYuan <yy19902439@126.com>

SamYuan1990 committed:
Fix for FAB-17151, FAB-17152, FAB-17149
Rename test folder name in test-env-%s- format and verfied clean up function works.

Signed-off-by: SamYuan <yy19902439@126.com>

guoger commented on core/ledger/kvledger/bookkeeping/test_exports.go:
```diff
@@ -33,7 +34,8 @@ type TestEnv struct {
 func NewTestEnv(t testing.TB) *TestEnv {
-	dbPath, err := ioutil.TempDir("", "bookkeep")
+	dbPath, err := ioutil.TempDir("", fmt.Sprintf("test-env-%s-", t.Name()))
+
```
kill empty line

guoger commented on core/ledger/kvledger/txmgmt/privacyenabledstate/test_exports.go:
```diff
@@ -48,7 +48,10 @@ type LevelDBCommonStorageTestEnv struct {
-	dbPath, err := ioutil.TempDir("", "cstestenv")
+	dbPath, err := ioutil.TempDir("", fmt.Sprintf("test-env-%s-", t.Name()))
+	env.provider = dbProvider
+	env.dbPath = dbPath
```
even though no real harm, this should come after err check

SamYuan1990 replied:
fixed via error check. :-)

guoger commented on core/ledger/kvledger/txmgmt/privacyenabledstate/test_exports.go:
```diff
@@ -118,11 +120,12 @@ func (env *CouchDBCommonStorageTestEnv) setupCouch() string {
 // Init implements corresponding function from interface TestEnv
 func (env *CouchDBCommonStorageTestEnv) Init(t testing.TB) {
-	redoPath, err := ioutil.TempDir("", "pestate")
+	redoPath, err := ioutil.TempDir("", fmt.Sprintf("test-env-%s-", t.Name()))
```
```suggestion
	redoPath, err := ioutil.TempDir("", fmt.Sprintf("bookkeeping-%s-", t.Name()))
```

SamYuan1990 replied:
https://github.com/hyperledger/fabric/pull/351/files#diff-4bd6ce0c76aa7b9fb651a7bdb242dd0eL36-R38
I think you mean those lines? for bookkeeping? in fact current file is used for couched or leveldb, if we want to naming it... "leveldbtest-env-%s" ?

SamYuan1990 replied:
resolve as out of date

SamYuan1990 committed:
Fix lint error

Signed-off-by: SamYuan <yy19902439@126.com>

guoger reviewed (changes requested):
thx for your contribution, some general comments to be considered:

- PR title should be descriptive, "Sam fix dec02" is not
- Commit Messages should follow [contribution guideline](https://hyperledger-fabric.readthedocs.io/en/latest/github/github.html) to include necessary info (they go into our git history)

```
You can now create your commit containing the changes you just added. Your commit message must contain the following information:
  - one line summary of the work in this commit as title, followed by an empty line
  - in the commit message body, explain why this change is needed, and how you approached it. This helps reviewers better understand your code and often speeds up the review process.
  - link to JIRA item or JIRA number, i.e. FAB-XXXXX
  - (optional) if no new tests are added, how the code is tested
```

SamYuan1990 committed:
Fix UT error

Signed-off-by: SamYuan <yy19902439@126.com>

SamYuan1990 committed:
Fix UT bug
Refactor UT with cleanup function to delete temp file

Signed-off-by: SamYuan <yy19902439@126.com>

jyellick commented:
@manish-sethi Could you take a look at this?

guoger commented on core/ledger/kvledger/txmgmt/privacyenabledstate/test_exports.go:
```diff
@@ -48,7 +49,8 @@ type LevelDBCommonStorageTestEnv struct {
 func (env *LevelDBCommonStorageTestEnv) Init(t testing.TB) {
-	dbPath, err := ioutil.TempDir("", "cstestenv")
+	dbPath, err := ioutil.TempDir("", strings.ReplaceAll(fmt.Sprintf("test-env-%s-", t.Name()), "/", "-"))
+	env.dbPath = dbPath
```
nit - should come after err check

guoger commented on core/ledger/kvledger/txmgmt/privacyenabledstate/test_exports.go:
```diff
@@ -118,11 +120,12 @@ func (env *CouchDBCommonStorageTestEnv) setupCouch() string {
 	}
-
+	env.redoPath = redoPath
+	logger.Infof(env.couchAddress)
```
do we need this in test?

guoger commented on core/ledger/kvledger/txmgmt/privacyenabledstate/test_exports.go:
```diff
@@ -153,10 +156,15 @@ func (env *CouchDBCommonStorageTestEnv) Init(t testing.TB) {
 		stateDBConfig,
 		[]string{"lscc", "_lifecycle"},
 	)
+	logger.Infof("processing")
```
do we need this?

also, as a general Go practice, err check should come immediately after where err might be produced. let's not insert code in between.

guoger commented on core/ledger/kvledger/txmgmt/privacyenabledstate/test_exports.go:
```diff
@@ -153,10 +156,15 @@ func (env *CouchDBCommonStorageTestEnv) Init(t testing.TB) {
 	assert.NoError(t, err)
 	env.t = t
 	env.provider = dbProvider
-	env.redoPath = redoPath
```
i'm not sure if `redoPath` is actually being used... @manish-sethi can we safely delete it from struct?

manish-sethi replied:
This is getting used in the cleanup function for deleting the redoPath dir

guoger commented on core/ledger/kvledger/txmgmt/privacyenabledstate/test_exports.go:
```diff
@@ -178,5 +186,6 @@ func (env *CouchDBCommonStorageTestEnv) Cleanup() {
 	os.RemoveAll(env.redoPath)
 	env.bookkeeperTestEnv.Cleanup()
 	env.provider.Close()
+	env.couchAddress = ""
```
this is not needed

SamYuan1990 replied:
This one is needed. I don't know how to... show git link here, but back to Init function 
	if env.couchAddress == "" {
		env.couchAddress = env.setupCouch()
	}
we should clean up this. otherwise, 
testenv.Init() will not create anymore, and throw up an error as db closed.

guoger reviewed (changes requested):
several feedbacks:
- it's generally not a good practice to combine several JIRA items into one PR, **unless** all of them are really trivial. I don't think this PR is the case, based on number of files and code it touches. Probably not worth splitting for now, but pls keep in mind next time :)
- err check should come immediately after where it's produced, w/o code in between
- i'm still seeing fairly amount of tmp dirs left after test:
```
bookkeeping-env-TestGetStateRangeScanIterator-couchDBCommonStorageTestEnv-068848688
bookkeeping-env-TestGetStateRangeScanIterator-levelDBCommonStorageTestEnv-669820366
bookkeeping-env-TestHandleChainCodeDeployOnCouchDB-couchDBCommonStorageTestEnv-729079094
bookkeeping-env-TestIterator-043434407
bookkeeping-env-TestIterator-231683142
bookkeeping-env-TestIterator-255042213
bookkeeping-env-TestIterator-503136104
bookkeeping-env-TestIterator-577301229
bookkeeping-env-TestIterator-694256858
bookkeeping-env-TestIterator-744094656
bookkeeping-env-TestIterator-801242537
bookkeeping-env-TestIterator-925915890
bookkeeping-env-TestIterator-977198623
bookkeeping-env-TestLongDBNameOnCouchDB-couchDBCommonStorageTestEnv-298196580
bookkeeping-env-TestQueryOnCouchDB-couchDBCommonStorageTestEnv-685071586
test-env-TestGetStateRangeScanIterator-couchDBCommonStorageTestEnv-111818965
test-env-TestGetStateRangeScanIterator-levelDBCommonStorageTestEnv-718851435
test-env-TestHandleChainCodeDeployOnCouchDB-couchDBCommonStorageTestEnv-859733107
test-env-TestLongDBNameOnCouchDB-couchDBCommonStorageTestEnv-683535577
test-env-TestQueryOnCouchDB-couchDBCommonStorageTestEnv-806973391
```


@manish-sethi i wonder why we need  `testEnvMaps`? we only have two structs there (leveldb/couchdb) and we do branch off based on different type. Wouldn't it be simpler if we just instantiate test env within those tests, instead of having a global map?

SamYuan1990 committed:
Remove unused log msg

Signed-off-by: SamYuan <yy19902439@126.com>

SamYuan1990 committed:
Fix new founding temp dir

Signed-off-by: SamYuan <yy19902439@126.com>

SamYuan1990 commented:
@guoger , thanks for new founding temp dirs.  It seems I missed -count=1 and local test used cache data.

SamYuan1990 committed:
Fix for TestIterator invoke clean up function

Signed-off-by: SamYuan <yy19902439@126.com>

guoger commented:
admittedly i wasn't looking closely enough at those tests, but i do think there are some substantial improvements needed here. It seems that tests individually initialize test env, and rely on `TestMain` to clean up. Therefore, we end up cleaning the last one only.

IIUC, we aim to avoid creating CouchDB instance for each test there, and i think we should restructure those tests to make this clear - having a single place to setup and teardown test env for a group of tests, who creates their own temp resources on demand and clear up by themselves.

But i'll defer to @manish-sethi to make the call

@SamYuan1990 i don't think `-count=1` and cache data has anything to do with those leftovers...

SamYuan1990 commented:
@guoger , Test with cache... it will skip some test when you rerun the test after changed some test code. see, we test once and find some folder(a,b,c), we updated some of them(a,b), clean up local folder as we are going to re run test and check folder again.... as test c has no changes, seems that go test will skip c, as it passed by cache. and folder c disappear at local, but we didn't fix test c.

guoger commented on core/ledger/kvledger/txmgmt/privacyenabledstate/db_test.go:
```diff
@@ -150,7 +151,7 @@ func TestGetStateMultipleKeys(t *testing.T) {
 	env.Init(t)
 	db := env.GetDBHandle(generateLedgerID(t))
-
+	defer env.Cleanup()
```
nit - `defer` should immediately follow `env.Init()` (especially when you are doing this at other places already)

SamYuan1990 replied:
completed.

guoger commented on core/ledger/kvledger/txmgmt/privacyenabledstate/test_exports.go:
```diff
@@ -118,11 +120,11 @@ func (env *CouchDBCommonStorageTestEnv) setupCouch() string {
 func (env *CouchDBCommonStorageTestEnv) Init(t testing.TB) {
-	redoPath, err := ioutil.TempDir("", "pestate")
+	redoPath, err := ioutil.TempDir("", strings.ReplaceAll(fmt.Sprintf("test-env-%s-", t.Name()), "/", "-"))
 	if err != nil {
```
not yours but could you do a follow-up to convert these to use `assert`? thx

SamYuan1990 replied:
replaced by 	assert.NoError(t, err)

guoger commented on core/ledger/kvledger/txmgmt/privacyenabledstate/test_exports.go:
```diff
@@ -153,10 +155,12 @@ func (env *CouchDBCommonStorageTestEnv) Init(t testing.TB) {
 		stateDBConfig,
 		[]string{"lscc", "_lifecycle"},
 	)
+	if err != nil {
```
kill this

guoger reviewed (changes requested):
I think current UT structure needs some rework, but this task is to remove leftover dirs. It generally looks good to me other than some nits.

Also, please do a rebase on master, and evaluate if we need to cherry-pick this to release-1.4 (checking if there are leftover dirs as well)

SamYuan1990 committed:
Update for code review.

Signed-off-by: SamYuan1990 <yy19902439@126.com>

SamYuan1990 committed:
Merge branch 'master' into SamFixDec02

SamYuan1990 committed:
Fix for FAB-17151, FAB-17152, FAB-17149

Rename test folder name in new format.
Updated clean up function cleaned temp dir created in test cases.
Updated test case to invoke cleanup function with defer to remove temp files.

Signed-off-by: SamYuan1990 <yy19902439@126.com>

SamYuan1990 commented:
@guoger , thanks for your review and comments. Just fixed conflicts with master. :-)

guoger commented on common/ledger/util/leveldbhelper/pkg_test.go:
```diff
@@ -49,12 +49,14 @@ func (dbEnv *testDBEnv) cleanup() {
 	if dbEnv.db != nil {
 		dbEnv.db.Close()
 	}
+	os.RemoveAll(dbEnv.path)
```
this duplicates what we have in `assert.NoError`

SamYuan1990 replied:
fixed.

guoger commented on common/ledger/util/leveldbhelper/pkg_test.go:
```diff
@@ -49,12 +49,14 @@ func (dbEnv *testDBEnv) cleanup() {
 	if providerEnv.provider != nil {
 		providerEnv.provider.Close()
 	}
+	os.RemoveAll(providerEnv.path)
```
this duplicates what we have in `assert.NoError`

guoger commented on core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/txmgr_test.go:
```diff
@@ -31,6 +31,7 @@ func TestTxSimulatorWithNoExistingData(t *testing.T) {
 		t.Logf("Running test for TestEnv = %s", testEnv.getName())
 		testLedgerID := "testtxsimulatorwithnoexistingdata"
 		testEnv.init(t, testLedgerID, nil)
+		defer testEnv.cleanup()
```
in general, `defer` in loop often leads to unexpected behavior and should be used with caution - in this case, you will be calling `cleanup` on the **last** `testEnv` multiple times

Either clean up w/o defer (like what we already have here), or assign testEnv to a new var and call defer.

guoger commented on core/ledger/kvledger/txmgmt/txmgr/lockbasedtxmgr/txmgr_test.go:
```diff
@@ -116,6 +117,7 @@ func TestTxSimulatorWithExistingData(t *testing.T) {
 		t.Run(testEnv.getName(), func(t *testing.T) {
 			testLedgerID := "testtxsimulatorwithexistingdata"
 			testEnv.init(t, testLedgerID, nil)
+			defer testEnv.cleanup()
```
ditto

guoger reviewed (changes requested):
Seems that leftover dirs are cleaned up after 0b5923bd9be7a17e5bd8c223b18916562826d40f but we could still accept changes on names of tmpdirs, so we have a better idea in the future about leaky tests.

Also, please squash and rebase on your next update so we have a clean history on merge.

SamYuan1990 commented:
Just update the code and I think I need help for  squash and rebase , as there was a merge with master, and now , if I try to merge the commit after merge with commits before merge as git rebase -i HEAD~4 ,it makes me confuse. If it's hard to merge commits across branch merge, I am ok to close this PR and do the changes in a new PR. @guoger

guoger commented:
@SamYuan1990 What you might wanna try is probably *rebase* instead of *merge* to resolve conflicts, but i'm fine if you open another PR. (description should be updated though, cuz it's no longer fixing leftovers)

manish-sethi commented:
> @manish-sethi Could you take a look at this?
@jyellick - I commented on Jira (https://jira.hyperledger.org/browse/FAB-17152). I think that this is perhaps already fixed in the master @DereckLuo will check. Also, he is making some other changes in these files related to this, so I would prefer to hold this PR and hear from @DereckLuo

manish-sethi commented:
> Seems that leftover dirs are cleaned up after [0b5923b](https://github.com/hyperledger/fabric/commit/0b5923bd9be7a17e5bd8c223b18916562826d40f) but we could still accept changes on names of tmpdirs, so we have a better idea in the future about leaky tests.
> 
> Also, please squash and rebase on your next update so we have a clean history on merge.

This clean up code was always there, then was removed in https://gerrit.hyperledger.org/r/#/c/fabric/+/31417/ and https://gerrit.hyperledger.org/r/#/c/fabric/+/31543/ and now finally added back again by @DereckLuo in the commit that you pointed out.

About the changes in the names of tmpdirs, yes sure we can take that change but this PR after rebase, has this cleanup calls twice.

manish-sethi commented:
> admittedly i wasn't looking closely enough at those tests, but i do think there are some substantial improvements needed here. It seems that tests individually initialize test env, and rely on `TestMain` to clean up. Therefore, we end up cleaning the last one only.
> 
> IIUC, we aim to avoid creating CouchDB instance for each test there, and i think we should restructure those tests to make this clear - having a single place to setup and teardown test env for a group of tests, who creates their own temp resources on demand and clear up by themselves.
> 
> But i'll defer to @manish-sethi to make the call
That's exactly what @DereckLuo is fixed under this https://jira.hyperledger.org/browse/FAB-17238. This still has some more clean up to do that he would be taking up under a separate Jira.

guoger commented:
> Seems that leftover dirs are cleaned up after [0b5923b](https://github.com/hyperledger/fabric/commit/0b5923bd9be7a17e5bd8c223b18916562826d40f) but we could still accept changes on names of tmpdirs, so we have a better idea in the future about leaky tests.
> 
> Also, please squash and rebase on your next update so we have a clean history on merge.

@manish-sethi right, as i replied to author, leftovers are cleaned up via other patches. We could still make tempdir names more descriptive as this PR currently stands. But author seems to have some problem rebasing it. Will close PR for now. @SamYuan1990 let me know if you are still interested in opening another PR.
